streams/src/main/java/org/apache/kafka/streams/state/TimestampedKeyValueStore.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state;
+
+/**
+ * A key-(value/timestamp) store that supports put/get/delete and range queries.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public interface TimestampedKeyValueStore<K, V> extends KeyValueStore<K, ValueAndTimestamp<V>> {
+
+    /**
+     * Update the value and timestamp associated with this key.
+     *
+     * @param key The key to associate the value to
+     * @param value The value to update, it can be {@code null};
+     *              if the serialized bytes are also {@code null} it is interpreted as deletes
+     * @param timestamp the timestamp to be stored next to the value; ignored if {@code value} is {@code null}
```

Comment:
Should we also state that the returned timestamp may also be `unknown` even though the value is not null? Ditto below.

Reply:
This would only happen during upgrade. Not sure if it is necessary?

Also, I don't plan to put any restrictions on the timestamp, ie, a PAPI use could put whatever she wants (including negative timestamps -- this would only fail if the store is logged because Kafka  Producer does not allow to set negative timestamps.

Thinking about this twice, I am wondering if we should add a check for negative timestamps? Note, that the `XxxStoreFaced` also uses `-1` as default timestamp on `put(key, value)` -- we could also use `0` as default though.

Thoughts? \cc @bbejeck @vvcephei @ableegoldman